Adding a readme for deepsparse.server (#285)
neuralmagic/deepsparse: README.md

limitations under the License.
-->

<h1><img alt="tool icon" src="https://raw.githubusercontent.com/neuralmagic/deepsparse/main/docs/source/icon-deepsparse.png" />&nbsp;&nbsp;DeepSparse</h1>

> <h3>Neural network inference engine that delivers GPU-class performance for sparsified models on CPUs</h3>

Reviewer: Original software tag lines need to remain for now as there is a separate review process for these, with several cross-department staff involved. They are used in multiple places so we want to change them relatively at once.

    </a>
</p>

> <h3> DeepSparse is a deep learning inference engine for running super-fast sparse models. 🚀🚀🚀</h3>

Reviewer: See previous comment about new tag lines.

> A CPU runtime engine delivering GPU-class performance by taking advantage of sparsity within neural networks to reduce compute required as well as accelerate memory bound workloads. Read more about sparsification [here](https://docs.neuralmagic.com/main/source/getstarted.html#sparsification).

Reviewer: Several cross-department staff members are involved with the short descriptions of software components. cc @NMsasa

The DeepSparse Engine is a CPU runtime that delivers GPU-class performance by taking advantage of sparsity (read more about sparsification [here](https://docs.neuralmagic.com/main/source/getstarted.html#sparsification)) within neural networks to reduce compute required as well as accelerate memory bound workloads. It is focused on model deployment and scaling machine learning pipelines, fitting seamlessly into your existing deployments as an inference backend.
A CPU runtime engine delivering GPU-class performance by taking advantage of sparsity within neural networks to reduce compute required as well as accelerate memory bound workloads. Read more about sparsification [here](https://docs.neuralmagic.com/main/source/getstarted.html#sparsification).

> The [GitHub repository](https://github.com/neuralmagic/deepsparse) includes package APIs along with examples to quickly get started benchmarking and inferencing sparse models.

Reviewer: We include this line because the page gets rendered in a few places outside of GitHub. Users may not see the GitHub repo first.

To begin your inference adventure with DeepSparse is super simple. You can either run DeepSparse in CLI or in Python. Let's first check out the CLI features... 👇

## DeepSparse Server & Benchmark in CLI

After you've installed DeepSparse in your Python environment, you have two resources via CLI at your disposal:

`deepsparse.server` : If you are interested in serving your models.

`deepsparse.benchmark` : If you are interested in experimenting with the throughput/latency performance of your models under different inference scenarios.

### 1. DeepSparse Server

The DeepSparse inference server allows you to serve models and pipelines in deployment. The server runs on top of the popular FastAPI web framework and Uvicorn web server.

 - run `deepsparse.server -h` to lookup arguments.

##### ⭐ Single Model ⭐

Example CLI command for running inference with a single model:

```bash
deepsparse.server \
    --task question_answering \
    --model_path "zoo:nlp/question_answering/bert-base/pytorch/huggingface/squad/base-none"
```

##### ⭐ Multiple Models ⭐
To serve multiple models in your deployment you can easily build a `config.yaml`. 

> In the example below, we are defining 2 BERT models in our configuration:

Reviewer: ```suggestion
In the example below, we are defining two BERT models in our configuration:
```

    models:
    - task: question_answering
        model_path: zoo:nlp/question_answering/bert-base/pytorch/huggingface/squad/base-none
        batch_size: 1
        alias: question_answering/dense
    - task: question_answering
        model_path: zoo:nlp/question_answering/bert-base/pytorch/huggingface/squad/pruned_quant-aggressive_95
        batch_size: 1
        alias: question_answering/sparse_quantized

Finally, after your `config.yaml` file is built, you can run the server with the config file path as an argument:
```bash
deepsparse.server --config_file config.yaml
```

### 2. DeepSparse Benchmark

The benchmark tool is available on your CLI to run expressive model benchmarks on the DeepSparse Engine with minimal parameters.

Run `deepsparse.benchmark -h` to look up arguments.

```shell
deepsparse.benchmark [-h] [-b BATCH_SIZE] [-shapes INPUT_SHAPES]
                          [-ncores NUM_CORES] [-s {async,sync}] [-t TIME]
                          [-nstreams NUM_STREAMS] [-pin {none,core,numa}]
                          [-q] [-x EXPORT_PATH]
                            model_path

```

> - [Getting Started with CLI Benchmarking](https://github.com/neuralmagic/deepsparse/tree/main/src/deepsparse/benchmark_model) Includes examples of inference scenarios such as: 

Reviewer: ```suggestion
- [Getting Started with CLI Benchmarking](https://github.com/neuralmagic/deepsparse/tree/main/src/deepsparse/benchmark_model) includes examples of inference scenarios such as: 
```

engine = compile_model(onnx_filepath, batch_size)
outputs = engine.run(inputs)
```
Compatibility/Support Notes
- ONNX version 1.5-1.7
- ONNX opset version 11+
- ONNX IR version has not been tested at this time

For a more in-depth read on available APIs and workflows, check out the [examples](https://github.com/neuralmagic/deepsparse/blob/main/examples/) and [DeepSparse Engine documentation.](https://docs.neuralmagic.com/deepsparse)

__ __

## Scheduling Single-Stream, Multi-Stream, and Elastic Inference ⚡

> The DeepSparse Engine offers up to 3 types of inferences based on your use-case. You can read more details here: [Inference Types](https://github.com/neuralmagic/deepsparse/blob/main/docs/source/scheduler.md).

Reviewer: ```suggestion
The DeepSparse Engine offers up to three types of inferences based on your use case. You can read more details here: [Inference Types](https://github.com/neuralmagic/deepsparse/blob/main/docs/source/scheduler.md).
```

1 ⚡ Single stream scheduling: the latency/synchronous scenario, requests execute serially. [`default`]

<img src="https://raw.githubusercontent.com/neuralmagic/deepsparse/main/docs/source/single-stream.png" alt="single stream diagram" />

PRO TIP: It's highly optimized for minimum per-request latency, using all of the system's resources provided to it on every request it gets.

2 ⚡ Multi-stream scheduling: the throughput/asynchronous scenario, requests execute in parallel.

<img src="https://raw.githubusercontent.com/neuralmagic/deepsparse/main/docs/source/multi-stream.png" alt="multi stream diagram" />

PRO TIP: The most common use cases for the multi-stream scheduler are where parallelism is low with respect to core count, and where requests need to be made asynchronously without time to batch them.

3 ⚡ Elastic scheduling: requests execute in parallel, but not multiplexed on individual NUMA Nodes.

PRO TIP: A workload that might benefit from the elastic scheduler is one in which multiple requests need to be handled simultaneously, but where performance is hindered when those requests have to share an L3 cache.
__ __

## Hardware Support 🧰

> The DeepSparse Engine is validated to work on x86 Intel and AMD CPUs running Linux operating systems. Mac and Windows require running Linux in a Docker or virtual machine.

Reviewer: We are updating these two paragraphs for AMD reasons. Updated text to follow.

> The DeepSparse Engine is validated to work on x86 Intel and AMD CPUs running Linux operating systems. Mac and Windows require running Linux in a Docker or virtual machine.

Reviewer: ```suggestion
The DeepSparse Engine is validated to work on x86 Intel (Haswell generation and later) and AMD CPUs running Linux, with support for AVX2, AVX-512, and VNNI instruction sets. Mac and Windows require running Linux in a Docker or virtual machine.
```

> It is highly recommended to run on a CPU with AVX-512 instructions available for optimal algorithms to be enabled. 

Reviewer: ```suggestion

```

## Resources 🔧

<table>
<tr><th> Documentation </th><th> &emsp;&emsp;&emsp;Versions </th><th> Info </th></tr>
<tr><td>

[DeepSparse](https://docs.neuralmagic.com/deepsparse/)

[SparseML](https://docs.neuralmagic.com/sparseml/)

[SparseZoo](https://docs.neuralmagic.com/sparsezoo/)

[Sparsify](https://docs.neuralmagic.com/sparsify/)

</td><td>

&emsp;stable : : [DeepSparse](https://pypi.org/project/deepsparse)

&emsp;nightly (dev) : : [DeepSparse-Nightly](https://pypi.org/project/deepsparse-nightly/)

&emsp;releases : : [GitHub](https://github.com/neuralmagic/deepsparse/releases)

</td><td>

[Blog](https://www.neuralmagic.com/blog/) 
[Resources](https://www.neuralmagic.com/resources/)

</td></tr> </table>

## Community 🧙

### Be Part of the Future... And the Future is Sparse!

Contribute with code, examples, integrations, and documentation as well as bug reports and feature requests! [Learn how here.](https://github.com/neuralmagic/deepsparse/blob/main/CONTRIBUTING.md)

### Join

> For user help or questions about DeepSparse, sign up or log in to our [Deep Sparse Community Slack](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ). We are growing the community member by member and happy to see you there. Bugs, feature requests, or additional questions can also be posted to our [GitHub Issue Queue.](https://github.com/neuralmagic/deepsparse/issues)

Reviewer: ```suggestion
For user help or questions about DeepSparse, sign up or log in to our [**Deep Sparse Community Slack**](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ). We are growing the community member by member and happy to see you there. Bugs, feature requests, or additional questions can also be posted to our [GitHub Issue Queue.](https://github.com/neuralmagic/deepsparse/issues)
```

Reviewer: GTM is purposely bolding the Slack Community CTA to draw more attention.
